src/common/libjob/spec.h

```diff
@@ -0,0 +1,137 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#ifndef _FLUX_CORE_SPEC_H
+#define _FLUX_CORE_SPEC_H
+
+typedef struct flux_jobspec1 flux_jobspec1_t;
+
+/* Remove the value in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator. Return 0 on success, -1 otherwise.
+ */
+int flux_jobspec1_attr_del (flux_jobspec1_t *jobspec, const char *path);
+
+/* Set the value in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator.
+ * 'fmt' should be a jansson pack-style string corresponding to the
+ * remaining args.
+ * Return 0 on success, -1 otherwise.
+ */
+int flux_jobspec1_attr_pack (flux_jobspec1_t *jobspec,
+                             const char *path,
+                             const char *fmt,
+                             ...);
+
+/* Unpack the values in the jobspec's 'attributes' section at the given path,
+ * where "." is treated as a path separator.
+ * 'fmt' should be a jansson unpack-style string corresponding to the
+ * remaining args.
+ * Return 0 on success, -1 otherwise.
+ */
+int flux_jobspec1_attr_unpack (flux_jobspec1_t *jobspec,
+                               const char *path,
+                               const char *fmt,
+                               ...);
+
+/* Check the validity of the 'attributes' section of a jobspec, writing an error
+ * message. Return 0 on success, -1 on error with an error message written.
+ * This function succeeding does not guarantee that the jobspec is valid.
+ */
+int flux_jobspec1_attr_check (flux_jobspec1_t *jobspec, char *errbuf, int errbufsz);
+
+/* Remove an entry in a Jobspec's environment.
+ * This function succeeds regardless of the presence of the variable.
+ * However, it will fail if the environment object does not exist.
+ * Return 0 on success, nonzero on error.
+ */
+int flux_jobspec1_env_unset (flux_jobspec1_t *jobspec, const char *name);
+
```

Comment:
So you mean return 0 in all circumstances? I am not sure I see the parallel with ``unsetenv`` because with ``unsetenv`` there is no case where there is no environment at all. But in any case I think that's reasonable behavior and if you think it's less surprising then I'll make the change.

Reply:
I must've been thinking that an empty environment might appear as a missing environment object, but since the environment object is created in the constructor, that can't really happen, cant it?

So I guess it's fine as is.